### ISSUE TITLE
Further Iteration - Type Stack-Dependent Checks

### ISSUE THREAD

**Issue by conrad-watt** [WebAssembly/relaxed-dead-code-validation#1]
**TL;DR**: Perform additional dead code checks over what was previously proposed. The original plan was to use the spec's division between "instruction syntax" and "validation", and to only perform instruction syntax checks on dead code. The new proposal is to perform any check that doesn't depend on popping from the type stack (all instruction syntax checks, and a subset of validation checks). This still allows us to avoid a polymorphic stack while addressing @rossberg's concerns about making instruction syntax observable.

-----------------

@rossberg made a [great comment](https://github.com/WebAssembly/design/issues/1379#issuecomment-701939042) summing up his concerns about the change. I'm going to be focussing mainly on his third bullet point here.

> Conceptually, the change makes the parsing/validation distinction observable. That takes away a relevant degree of freedom from both spec and implementations. And ensuring that impls get it right will require a lot of subtle tests, even more so than under the current semantics. (Related, some changes we did in proposals would technically become breaking changes under this semantics, e.g., reinterpreting the alignment immediate of load/store instructions as a bit field in the multi memory proposal.)

**Scene-setting**

First - it's become clear that different people in the original issue meant different things by "validation". The formal spec defines two levels of restriction on the structure of instruction sequences. First, the [instruction syntax](https://webassembly.github.io/spec/core/binary/instructions.html) defines the bytecode mapping for each abstract operation, including their immediate parameters. For example, take `i32.load`. Its (slightly simplified) instruction syntax is:
```
0x28  (a : u32) (o : u32)
```
where `u32` is a range-restricted LEB. Just from this definition, `i32.load (a=2^40) (o=2^40)` is known to be forbidden. Even though `2^40` can be legally expressed by WebAssembly's LEB encoding, it is beyond the bounds of a `u32`.

Separately, the specification defines validation rules over a parsed AST. The (simplified) validation rule for `i32.load` is
```
C.memory[0] = m    2^a <= bytesize(i32)
---------------------------------------------
C ⊢ i32.load a o : [i32]→[i32]
```
From this (plus validation of the module/function, which defines `C`), we know that the instruction `i32.load (a=50) (o=0)`
is forbidden, since even though `50` is a `u32`, it is outside the allowed bounds. Note that `o` is _only_ restricted by the instruction syntax, not any constraint of the validation rule.

It is an explicit design goal of the language that it should be possible to carry out these checks in a streaming manner. Instead of fully constructing the AST before applying validation, a regular implementation is expected to carry out both instruction syntax and validation checks simultaneously on a single instruction before moving on to the next. So a concrete implementation might look something like
```
decode_and_validate_instruction_sequence(...) {
...
  while ( nextOp = stream.getOp() )
    decode_and_validate_instruction(nextOp);
}

decode_and_validate_instruction(typeContext, nextOp) {
  switch (nextOp)
  ...
  | i32.load =>
      a = stream.getU32();
      check ( a <= 2 );
      o = stream.getU32();
      typeStack.pop ( i32 );
      typeStack.push( i32 );
      ...
}
```
Here, the instruction syntax checks on the immediates are carried out as the LEB is read, which is a common strategy in implementations.

Some people in the initial issue used "validation" to refer to the entire process of processing a single instruction in this way, which incorporates both the instruction syntax checks, and the (spec) validation checks.

In our initial proposal, I used the spec's distinction between instruction syntax checks and validation checks, proposing that dead code should _only_ undergo instruction syntax checks. Imagine I made a bool-templated version of `decode_and_validate_instruction` as follows:
```
decode_and_validate_instruction<doValidate>(typeContext, nextOp) {
  switch (nextOp)
  ...
  | i32.load =>
      a = stream.getU32();
      #if doValidate
          check ( a <= 2 );
      #endif
      o = stream.getU32();
      #if doValidate
          typeStack.pop ( i32 );
          typeStack.push( i32 );
      #endif
      ...
}
```

@rossberg made a great point that this distinction exists only in the spec, and we've exploited validation checks to make backwards-compatible changes to the instruction syntax (as described [here](https://github.com/WebAssembly/design/issues/1379#issuecomment-702432482) for multi-memory load). Implementers just do all the checks at once, give the same (class of) error message if any are violated, and consider them all to be under the umbrella of "validation".

The current polymorphic type stack is necessary to handle dead code following after a non-deterministically chosen type stack caused by the validation of an instruction such as `unreachable`:
```
C ⊢ unreachable : t1*→t2*
```

The original proposal, in getting rid of (spec) validation for dead code, gets rid of the polymorphic stack. However, it does have the downside that checks on immediates in implementations would be forced to conform to the spec's pre-ordained instruction syntax vs validation split. This would leave us less able to perform encoding tricks such as the multi-memory load one described above.

**Draw the line closer to the expectations of implementations**

My revised proposal is to make a distinction between "type stack-dependent" checks, and "type stack-independent" checks. From the spec point of view, the "type stack-dependent" checks are a subset of the validation checks. In the revised proposal, dead code will only undergo all type stack-independent checks. The type stack-dependent checks are any check which involves a `pop` or a `push` to the type stack, or any check on a `popped` type (see @titzer's [comment ](https://github.com/WebAssembly/design/issues/1379#issuecomment-702725791) on how`return_call_ref` must be handled).

I believe this revised approach corresponds to @titzer's conceptualisation of "validation of immediates". Concretely, the above example for `i32.load` would become
```
decode_and_validate_instruction<doValidate>(typeContext, nextOp) {
  switch (nextOp)
  ...
  | i32.load =>
      a = stream.getU32();
      check ( a <= 2 );
      o = stream.getU32();
      #if doValidate
          typeStack.pop ( i32 );
          typeStack.push( i32 );
      #endif
      ...
}
```
Note that the spec validation check on `a` is now carried out.

**Upsides**

This preserves our ability to wiggle between instruction syntax and validation checks while still allowing the type stack to remain concrete. I don't believe that this approach is any harder to implement than the previous proposal (although please share your opinions).

**Downsides**

This will be a more invasive change to the formal spec. I don't believe there will be any ambiguity in the behaviour categories this revised proposal intends to describe (i.e. what is type stack-dependent), but the current formalisation of validation will need to change in order to express them aesthetically. There are several possible approaches. If we ever move this revised proposal to stage 2 (indicating consensus on the informal direction), when updating the formal spec I'll mechanise the revised type system to ensure that the type soundness proof and archetypal type checker are still correct.

**A note on existing behaviour**

There has been some concern raised that implementations will struggle to align on subtle changes to decoding and validation. From talking to people and running my own tests, it's clear that implementations already diverge in some areas. For example, JSC has actually never implemented the current polymorphic stack approach, so already admits programs which are in violation of the spec. Another subtle issue I found was that SpiderMonkey appears to read the alignment immediate `a` of `load` [as a fixed single-byte](https://searchfox.org/mozilla-central/source/js/src/wasm/WasmOpIter.h#1357), even though it may be encoded as an LEB with trailing zero-bytes. I haven't tested V8 yet.

Committing to this proposal means committing to some kind of one-off refactoring to each engine's decode-validate logic anyway. If we can conjure up a rigorous enough test suite, this may be an opportunity to bring the behaviour of engines closer together. Maybe in the long-term we need a more exhaustive conformance test suite that also stresses the binary encoding.

**Comment by RossTate** [comment]
I believe these are the only changes to the spec that need to be made for this:
1. Introduce a metavariable to represent stack types and change stack-polymorphic rules to use this metavariable. (This just makes the presentation more extensible and itself has no effect on validation.)
2. Add an unreachable state type to the grammar for that metavariable. (On its own, this also has no effect on validation.)
3. Add a rule that says an instruction can have the unreachable input state type and an arbitrary output state type if there exist input and output state types that would make the instruction type-check. (This single rule is the only real change to validation.)

If so, this should not be a particularly invasive change besides the fact that the first step has to rephrase existing rules to make the spec more extensible to being with.

**Comment by conrad-watt** [comment]
Let's continue the convo in this thread, but see my comment [here ](https://github.com/WebAssembly/design/issues/1379#issuecomment-702799887) in response. Similar issues occur with the pop-abstraction-breaking instructions @titzer describes [here](https://github.com/WebAssembly/design/issues/1379#issuecomment-702725791) (actually, his are probably better examples).

Really the issue is that the formal typing rules are trying to layer an abstraction over the true desired type checking algorithm, and we now want to break an abstraction in a way that makes sense for the type checking algorithm, but that the formal typing rules were previously relying on.

EDIT: I don't think the changes to the formalism are in any way insurmountable, but I do want to acknowledge the additional changes (to the formalism) that would be needed especially for post-MVP features.

**Comment by RossTate** [comment]
> This is enough to deal with the MVP instructions, but something more is needed to deal with post-MVP instructions such as (rtt.sub <type> (<rtt>)), where the comparison between its immediate type parameter and its stack operand would only be carried out if the input stack is not unreachable.

`rtt.sub n t1 t2` is valid in the unreachable state type according to the rule I gave iff `t1 <: t2` holds: a concrete state type you can use to prove this is `[(rtt n t1)]`.

`return_ref_call` is always valid in the unreachable state type according to the rule I gave: a concrete state type you can use to prove this is `[(ref (func (param) (result t1 ... tn)))]` where `t1 ... tn` matches the result type of the enclosing function.

**Comment by conrad-watt** [comment]
I guess that approach could work if, as a general principle, we ensure that (the formal) types on the type stack do not contain any references to the contents of the type context C, but always inline things like function type signatures. I believe the typed function references proposal currently formalises things using indices, so this approach wouldn't work without adjustment.
EDIT: another approach might be to use a bottom value type

In any case, this is purely a detail of the formalisation which is totally solvable one way or the other (it shouldn't affect the concrete representation of types in implementations), so I'm mainly interested in building consensus on the informal approach.
